sns/ut/isal_ub.c

```diff
@@ -0,0 +1,276 @@
+    if (gf_invert_matrix(b, invert_matrix, k) < 0)
+        return -1;
+
+    // Get decode matrix with only wanted recovery rows
+    for (i = 0; i < nerrs; i++) {
+        if (frag_err_list[i] < k)    // A src err
+            for (j = 0; j < k; j++)
+                decode_matrix[k * i + j] =
+                    invert_matrix[k * frag_err_list[i] + j];
+    }
+
+    // For non-src (parity) erasures need to multiply encode matrix * invert
+    for (p = 0; p < nerrs; p++) {
+        if (frag_err_list[p] >= k) {    // A parity err
+            for (i = 0; i < k; i++) {
+                s = 0;
+                for (j = 0; j < k; j++)
+                    s ^= gf_mul(invert_matrix[j * k + i],
+                            encode_matrix[k * frag_err_list[p] + j]);
+                decode_matrix[k * p + i] = s;
+            }
+        }
+    }
+    return 0;
+}
+
+static int ub_test(uint32_t k, uint32_t p, uint32_t len)
+{
+    uint32_t i, j, m;
+    uint32_t nerrs = 0;
+
+    // Fragment buffer pointers
+    uint8_t *frag_ptrs[MMAX];
+    uint8_t *recover_srcs[KMAX];
+    uint8_t *recover_outp[KMAX];
+    uint8_t frag_err_list[MMAX];
+
+    // Coefficient matrices
+    uint8_t *encode_matrix, *decode_matrix;
+    uint8_t *invert_matrix, *temp_matrix;
+    uint8_t *g_tbls;
+    uint8_t decode_index[MMAX];
+
+    m = k + p;
+
+    // Allocate coding matrices
+    encode_matrix = malloc(m * k);
+    decode_matrix = malloc(m * k);
+    invert_matrix = malloc(m * k);
+    temp_matrix = malloc(m * k);
+    g_tbls = malloc(k * p * 32);
+
+    if (encode_matrix == NULL || decode_matrix == NULL
+        || invert_matrix == NULL || temp_matrix == NULL || g_tbls == NULL) {
+        printf("Test failure! Error with malloc\n");
+        return -1;
+    }
+
+    // Allocate the src & parity buffers
+    for (i = 0; i < m; i++) {
+        if (NULL == (frag_ptrs[i] = malloc(len))) {
+            printf("alloc error: Fail\n");
+            return -1;
+        }
+    }
+
+    // Allocate buffers for recovered data
+    for (i = 0; i < p; i++) {
+        if (NULL == (recover_outp[i] = malloc(len))) {
+            printf("alloc error: Fail\n");
+            return -1;
+        }
+    }
+
+    // Fill sources with random data
+    for (i = 0; i < k; i++)
+        for (j = 0; j < len; j++)
+            frag_ptrs[i][j] = rand();
+
+    gf_gen_cauchy1_matrix(encode_matrix, m, k);
+
+    // Initialize g_tbls from encode matrix
+    ec_init_tables(k, p, &encode_matrix[k * k], g_tbls);
+
+    // Generate EC parity blocks from sources
+    ec_encode_data(len, k, p, g_tbls, frag_ptrs, &frag_ptrs[k]);
+
+    for (i = 0; i < p; i++)
+    {
+        uint32_t err_id = rand() % m;
+        frag_err_list[nerrs++] = err_id;
+        memset(frag_ptrs[err_id], 0xFF, len);
+    }
+
+    // Find a decode matrix to regenerate all erasures from remaining frags
+    gf_gen_decode_matrix_simple(encode_matrix, decode_matrix,
+                        invert_matrix, temp_matrix, decode_index,
```

Comment:
`invert_matrix` and `temp_matrix `  are only used inside this function.
We can allocate them inside the function `gf_gen_decode_matrix_simple()`.
It is important to reduce the scope of these variables.

Reply:
Done.